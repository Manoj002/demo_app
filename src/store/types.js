export const USER_REGISTERED = 'USER_REGISTERED';
export const SUBMIT_NAME = 'SUBMIT_NAME';
export const NAME_CHANGED = 'NAME_CHANGED';
