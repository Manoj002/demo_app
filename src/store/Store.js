import React, {createContext, useReducer} from 'react';
import {USER_REGISTERED, NAME_CHANGED, SUBMIT_NAME} from './types';

const initialState = {
  isUserRegistered: false,
  userName: undefined,
};

export const StoreContext = createContext(initialState);

export const reducer = (state, action) => {
  switch (action.type) {
    case USER_REGISTERED:
      return {...state, isUserRegistered: action.payload};

    case NAME_CHANGED:
      return {...state, userName: action.payload};

    case SUBMIT_NAME:
      return {...state, userName: action.payload};

    default:
      return state;
  }
};

export const StoreProvider = ({children}) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <StoreContext.Provider value={[state, dispatch]}>
      {children}
    </StoreContext.Provider>
  );
};
