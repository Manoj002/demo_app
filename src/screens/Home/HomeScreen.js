import React, {useContext, useState, useEffect} from 'react';
import {StatusBar, ToastAndroid, NativeModules, Platform} from 'react-native';
import {Text, Button} from '../../commonComponents';
import {NAME_CHANGED, USER_REGISTERED, SUBMIT_NAME} from '../../store/types';
import PropTypes from 'prop-types';
import {
  Container,
  ContentContainer,
  StyledTextInput,
  ButtonWrapper,
} from './HomeScreen.styles';
import {Colors} from '../../constants';
import {StoreContext} from '../../store/Store';
import Strings from '../../constants/Strings';

const HomeScreen = ({navigation}) => {
  const [isEmulator, setIsEmulator] = useState('');

  const checkIsEmulator = async () => {
    const {DeviceDetails} = NativeModules;
    const emulator = await DeviceDetails.isEmulator();
    setIsEmulator(emulator);
  };

  useEffect(() => {
    checkIsEmulator();
  }, []);

  const [{isUserRegistered, userName}, dispatch] = useContext(StoreContext);

  const renderToastMessage = displayMessage => {
    ToastAndroid.show(
      displayMessage
        ? `${displayMessage}`
        : 'Please Register, by entering your name above',
      ToastAndroid.SHORT,
    );
  };

  const handleNavigation = route => {
    if (isUserRegistered) {
      navigation.navigate(route);
    } else {
      renderToastMessage();
    }
  };

  return (
    <Container>
      <StatusBar animated backgroundColor={Colors.darkGray} />
      <Text marginLeft={20} marginBottom={24} marginTop={40} fontSize={18} bold>
        {Strings.WELCOME_USER},
      </Text>
      <StyledTextInput
        autoCapitalize="words"
        placeholder={Strings.PLACEHOLDER}
        placeholderTextColor={Colors.lightGray}
        testID="TextInput-UserName"
        onChangeText={changedUserName => {
          dispatch({type: NAME_CHANGED, payload: changedUserName});
        }}
        value={userName}
      />
      <ButtonWrapper>
        <Button
          colored
          marginBottom={20}
          marginTop={20}
          testID="Submit-btn"
          onPress={() => {
            if (userName) {
              dispatch({type: SUBMIT_NAME, payload: userName});
              dispatch({type: USER_REGISTERED, payload: true});
              renderToastMessage('User Registration Successful!');
            } else {
              renderToastMessage();
            }
          }}>
          {Strings.SUBMIT_NAME}
        </Button>
      </ButtonWrapper>
      <Text marginBottom={20} align="center" testID="display-label">
        {Strings.RUNNING_ON}{' '}
        {isEmulator
          ? Platform.OS === 'android'
            ? 'Emulator'
            : 'Simulator'
          : 'Device'}
      </Text>
      <ContentContainer>
        <Button
          testID="Account-BTN"
          buttonAsText
          onPress={() => handleNavigation('Account')}>
          {Strings.NAV_ACCOUNT}
        </Button>
        <Button buttonAsText onPress={() => handleNavigation('Profile')}>
          {Strings.NAV_PROFILE}
        </Button>
      </ContentContainer>
    </Container>
  );
};

HomeScreen.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func,
  }).isRequired,
};

export default HomeScreen;
