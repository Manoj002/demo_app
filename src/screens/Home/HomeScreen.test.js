import React from 'react';
import {ToastAndroid, NativeModules} from 'react-native';
import {render, fireEvent} from '@testing-library/react-native';
import HomeScreen from './HomeScreen';
import {StoreContext, StoreProvider} from '../../store/Store';

const navigation = {
  navigate: jest.fn(),
};

const mockValues = {
  isUserRegistered: true,
  userName: 'Manoj',
};

const mockDispatch = jest.fn();

describe('Home Screen', () => {
  beforeEach(() => {
    const isEmulator = jest.fn(() => Promise.resolve(true));
    NativeModules.DeviceDetails = {
      isEmulator,
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the Screen correctly', () => {
    const Container = render(
      <StoreProvider>
        <StoreContext.Provider value={[mockValues, mockDispatch]}>
          <HomeScreen navigation={navigation} />
        </StoreContext.Provider>
      </StoreProvider>,
    );
    expect(Container).toMatchSnapshot();
  });

  it('should display entered username', () => {
    const Container = render(
      <StoreProvider>
        <StoreContext.Provider value={[mockValues, mockDispatch]}>
          <HomeScreen navigation={navigation} />
        </StoreContext.Provider>
      </StoreProvider>,
    );
    const {getByTestId} = Container;
    const TextInput = getByTestId('TextInput-UserName');
    fireEvent.changeText(TextInput, 'Manoj');
    expect(TextInput.props.value).toBe('Manoj');
  });

  it('should display toast message when user hit submit button without entering username', () => {
    jest.spyOn(ToastAndroid, 'show');
    const Container = render(
      <StoreProvider>
        <StoreContext.Provider
          value={[
            {
              isUserRegistered: false,
              userName: undefined,
            },
            mockDispatch,
          ]}>
          <HomeScreen navigation={navigation} />
        </StoreContext.Provider>
      </StoreProvider>,
    );
    const {getByTestId} = Container;
    const SubmitButton = getByTestId('Submit-btn');
    fireEvent.press(SubmitButton);
    expect(ToastAndroid.show).toHaveBeenCalledWith(
      'Please Register, by entering your name above',
      ToastAndroid.SHORT,
    );
  });

  it('should register user on button press', async () => {
    const Container = render(
      <StoreProvider>
        <StoreContext.Provider value={[mockValues, mockDispatch]}>
          <HomeScreen navigation={navigation} />
        </StoreContext.Provider>
      </StoreProvider>,
    );
    const {getByTestId} = Container;
    const SubmitButton = getByTestId('Submit-btn');
    fireEvent.press(SubmitButton);
    expect(mockDispatch).toHaveBeenCalledTimes(2);
  });

  it('should navigate to AccountScreen on Button Press', () => {
    const Container = render(
      <StoreProvider>
        <StoreContext.Provider value={[mockValues, mockDispatch]}>
          <HomeScreen navigation={navigation} />
        </StoreContext.Provider>
      </StoreProvider>,
    );
    const {getByTestId} = Container;
    const AccountBTN = getByTestId('Account-BTN');
    fireEvent.press(AccountBTN);
    expect(navigation.navigate).toHaveBeenCalledWith('Account');
  });

  it('Should able to call native bridge and display when application is opened in device', async () => {
    const Container = render(
      <StoreProvider>
        <StoreContext.Provider value={[mockValues, mockDispatch]}>
          <HomeScreen navigation={navigation} />
        </StoreContext.Provider>
      </StoreProvider>,
    );
    const {getByTestId} = Container;
    const label = getByTestId('display-label');
    expect(label.props.children).toStrictEqual(['Running on', ' ', 'Devices']);
  });
});
