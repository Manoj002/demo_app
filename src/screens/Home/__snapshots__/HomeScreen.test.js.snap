// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen renders the Screen correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#212124",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <Text
    align="left"
    bold={true}
    color="lightGray"
    fontSize={18}
    lineHeight={18}
    marginBottom={24}
    marginLeft={20}
    marginTop={40}
    style={
      Array [
        Object {
          "color": "#818181",
          "fontFamily": "Roboto",
          "fontSize": 18,
          "fontWeight": "bold",
          "lineHeight": 18,
          "marginBottom": 24,
          "marginLeft": 20,
          "marginTop": 40,
          "textAlign": "left",
        },
      ]
    }
  >
    Welcome User
    ,
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="words"
    onChangeText={[Function]}
    placeholder="Enter Your Name"
    placeholderTextColor="#818181"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "borderColor": "#818181",
          "borderRadius": 4,
          "borderWidth": 2,
          "color": "#818181",
          "fontSize": 16,
          "marginHorizontal": 20,
          "paddingHorizontal": 16,
        },
      ]
    }
    testID="TextInput-UserName"
    underlineColorAndroid="transparent"
    value="Manoj"
  />
  <View
    style={
      Array [
        Object {
          "marginHorizontal": 20,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#007AFF",
          "borderRadius": 8,
          "height": 50,
          "justifyContent": "center",
          "marginBottom": 20,
          "marginTop": 20,
          "opacity": 1,
          "width": "100%",
        }
      }
      testID="Submit-btn"
    >
      <Text
        align="left"
        bold={false}
        color="white"
        fontSize={16}
        lineHeight={18}
        marginBottom={0}
        marginLeft={0}
        marginTop={0}
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Roboto",
              "fontSize": 16,
              "lineHeight": 18,
              "textAlign": "left",
            },
          ]
        }
      >
        Submit Name
      </Text>
    </View>
  </View>
  <Text
    align="center"
    bold={false}
    color="lightGray"
    fontSize={16}
    lineHeight={18}
    marginBottom={20}
    marginLeft={0}
    marginTop={0}
    style={
      Array [
        Object {
          "color": "#818181",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "lineHeight": 18,
          "marginBottom": 20,
          "textAlign": "center",
        },
      ]
    }
    testID="display-label"
  >
    Running on
     
    Devices
  </Text>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 20,
          "paddingVertical": 20,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "borderRadius": 8,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
      testID="Account-BTN"
    >
      <Text
        align="left"
        bold={false}
        color="blue"
        fontSize={16}
        lineHeight={18}
        marginBottom={0}
        marginLeft={0}
        marginTop={0}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontFamily": "Roboto",
              "fontSize": 16,
              "lineHeight": 18,
              "textAlign": "left",
            },
          ]
        }
      >
        Go to Account
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "borderRadius": 8,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        align="left"
        bold={false}
        color="blue"
        fontSize={16}
        lineHeight={18}
        marginBottom={0}
        marginLeft={0}
        marginTop={0}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontFamily": "Roboto",
              "fontSize": 16,
              "lineHeight": 18,
              "textAlign": "left",
            },
          ]
        }
      >
        Go to Profile
      </Text>
    </View>
  </View>
</View>
`;
